narwhal/primary/src/header_waiter.rs

```diff
@@ -291,49 +275,20 @@ impl HeaderWaiter {
                         Ok(header) => header,
                     };
                     if let Some(header) = header {
-                        let _ = self.pending.remove(&header.id);
+                        if let Some((_, tx_cancel)) = self.pending.remove(&header.id) {
+                            let _ = tx_cancel.send(());
+                        }
                         for x in header.payload.keys() {
                             let _ = self.batch_requests.remove(x);
                         }
                         for x in &header.parents {
                             let _ = self.parent_requests.remove(x);
                         }
-                        permit.send(header);
+                        // Ok to drop the header if core is overloaded.
+                        let _ = self.tx_headers_loopback.try_send(header);
```

Comment:
Yes, if the range sync ensures progress, but see above some corner cases that might need to be considered?

Reply:
Similarly discussed above.

Reply:
I am thinking again and again if that's ok to do. The thing is that right now we don't differentiate between headers that:
(1) are part of a certificate (hence already signed and we don't really care to sign to "help" nodes make progress)
(2) new headers that we haven't voted yet and we want/need to sign

I believe - and maybe as a more general design/approach - we should give a different priority to headers that are not signed so we can help the protocol advance (and ensure we'll not end up in some weird deadlock situation). In the end of the day though I guess at least `2f+1` nodes should be in a position where they have dropped the latest headers to not make the protocol advance. Anyways if we think this will not be an issue I am totally ok with this approach (one less reason to block!).

Reply:
IIUC the network model is that network can randomly drop any message, but if retried enough times eventually a message will go through. To the node proposing headers, it _should_ handle the case when messages to some / all other nodes get dropped, and it _should_ just resend the message. Also it makes no difference to the proposing node whether the message was dropped by the network or dropped by another node before it voted.

Basically, the liveness of proposing headers and voting should be based on the proposing node managing the state, votes and retries correctly. It should not require any particular message getting through the network and not get dropped by the other node.

Reply:
> To the node proposing headers, it should handle the case when messages to some / all other nodes get dropped, and it should just resend the message. Also it makes no difference to the proposing node whether the message was dropped by the network or dropped by another node before it voted.

The proposer is broadcasting the header to the other nodes and is retrying to each individual node until the header has been successfully received via network - we don't send back any acknowledgement that we successfully processed/stored the header. The proposer doesn't really know what happens after that point. So semantically there is a difference between a message has been dropped by network  Vs dropped by some internal process after it has been successfully received via network